tempesta_fw/sched/tfw_sched_rr.c

```diff
@@ -54,92 +138,353 @@ typedef struct {
+	for (idx = c_pidx; idx < c_max_idx; idx++) {
+		rcu_assign_pointer(s_entry->c_pentry[idx],
+				   s_entry->c_pentry[idx + 1]);
+		s_entry->c_pentry[idx]->pidx--;
+	}
+
+	/*
+	 * If this is the last connection to server, then make the server
+	 * unavailable to scheduler as well. Initialize the server entry
+	 * for next use. Move pointers to other RR server entries to make
+	 * the array of pointers contiguous without gaps.
+	 * 
+	 * The way the update is done may result in that some servers get
+	 * chosen twice in succession.
+	 */
+	if (atomic_dec_and_test(&s_entry->n_conn)) {
+		s_entry->srv = NULL;
+		s_entry->pidx = 0;
+		atomic64_set(&s_entry->rr_cnt, 0);
+
+		s_max_idx = atomic_read(&sched_data->n_srv) - 1;
+		for (idx = s_pidx; idx < s_max_idx; idx++) {
+			rcu_assign_pointer(sched_data->s_pentry[idx],
+					   sched_data->s_pentry[idx + 1]);
+			sched_data->s_pentry[idx]->pidx--;
+		}
+		if (atomic_dec_and_test(&sched_data->n_srv))
+			atomic64_set(&sched_data->rr_cnt, 0);
+		rcu_assign_pointer(sched_data->s_pentry[s_max_idx], NULL);
+	}
+
+	/*
+	 * Do this only after @n_conn is decremented. Some connections
+	 * may get chosen twice in succession, but that's better than
+	 * getting a NULL pointer.
+	 */
+	rcu_assign_pointer(s_entry->c_pentry[c_max_idx], NULL);
+	c_entry->pidx = 0;
+
+	/* This also serves as memory barrier for atomic set. */
+	spin_unlock(&sched_data->slock);
+
+	/*
+	 * Finally, reclaim the connection entry.
+	 *
+	 * Note that server entry is not reclaimed explicitly when it is
+	 * made unavailable to scheduler. That can't be done when there's
+	 * only TFW_SG_MAX_SRV server entries. The entry may get reused
+	 * immediately for a new connection, and we can't delay that
+	 * as entries are preallocated and assigned on permanent basis.
+	 * So a server entry may get used while it's still in somewhat
+	 * inconsistent state.
+	 *
+	 * Server is made unavailable to scheduler only together with
+	 * the last connection to that server. Pointers to connection
+	 * and server entries are both RCU-assigned, and they are both
+	 * RCU-dereferenced in the same RCU critical section. So when
+	 * the last connection entry is not accessed anymore and can
+	 * be reclaimed, the corresponding server entry has no users
+	 * as well. So a server entry can never lead to an invalid
+	 * connection entry. It may get reused immediately though,
+	 * so it's filled with data that point to a live connection.
+	 */
+	call_rcu(&c_entry->rcu, tfw_sched_rr_conn_reclaim);
+}
+
+static void
+tfw_sched_rr_update_data(TfwSrvGroup *sg, TfwSrvConnection *s_conn)
+{
+	TfwRrSchedData *sched_data;
+
+	BUG_ON(sg == NULL);
+	BUG_ON(s_conn == NULL);
+
+	rcu_read_lock();
+
+	sched_data = rcu_dereference(sg->sched_data);
+	BUG_ON(sched_data == NULL);
+
+	if (tfw_connection_live(&s_conn->conn))
+		tfw_sched_rr_go_live(sched_data, s_conn);
+	else
+		tfw_sched_rr_go_dead(sched_data, s_conn);
+
+	rcu_read_unlock();
```

Comment:
Wow... The things going there look so complex.... I believe the redesign is required, basicaly RR is the most simple thing and should not be so complex.